### ISSUE TITLE
[JENKINS-45277] Powershell Pipeline Step Hangs

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/durable-task-plugin#390]
<p><del>When running powershell commands in parallel the stages will hang. They say <span class="error">&#91;running powershell script&#93;</span> but nothing runs, I've tried different code snippets so it doesn't appear to be related to what I'm running</del></p>

<p>Powershell pipeline task just hangs regardless of running in parallel or not. I'm running Powershell 2.0 on Windows 7</p>

<p>Job attached</p>

<pre>
<span class="code-comment">//Source: https://github.com/jenkinsci/pipeline-examples/blob/master/pipeline-examples/parallel-from-list/parallelFromList.groovy
</span>
<span class="code-comment">// While you can<span class="code-quote">'t use Groovy'</span>s .collect or similar methods currently, you can
</span><span class="code-comment">// still transform a list into a set of actual build steps to be executed in
</span><span class="code-comment">// parallel.
</span>
<span class="code-comment">// Our initial list of strings we want to echo in parallel
</span>def computers = [<span class="code-quote">"COMPUTERNAME"</span>]

env.computers = computers

<span class="code-comment">// The map we'll store the parallel steps in before executing them.
</span>def stepsForParallel = [:]

<span class="code-comment">// The standard <span class="code-quote">'<span class="code-keyword">for</span> (<span class="code-object">String</span> s: stringsToEcho)'</span> syntax also doesn't work, so we
</span><span class="code-comment">// need to use old school <span class="code-quote">'<span class="code-keyword">for</span> (<span class="code-object">int</span> i = 0...)'</span> style <span class="code-keyword">for</span> loops.
</span><span class="code-keyword">for</span> (<span class="code-object">int</span> i = 0; i < computers.size(); i++) {
<span class="code-comment">// Get the actual string here.
</span>def s = computers.get(i)

<span class="code-comment">// Transform that into a step and add the step to the map as the value, with
</span><span class="code-comment">// a name <span class="code-keyword">for</span> the parallel step as the key. Here, we'll just use something
</span><span class="code-comment">// like <span class="code-quote">"echoing (string)"</span>
</span>def stepName = <span class="code-quote">"Pushing to ${s}"</span>

stepsForParallel[stepName] = transformIntoStep(stepName, i)
}

<span class="code-comment">// Actually run the steps in parallel - parallel takes a map as an argument,
</span><span class="code-comment">// hence the above.
</span>node{
parallel stepsForParallel
}

<span class="code-comment">// Take the string and echo it.
</span>def transformIntoStep(stepName, inputIndex) {
<span class="code-comment">// We need to wrap what we <span class="code-keyword">return</span> in a Groovy closure, or <span class="code-keyword">else</span> it's invoked
</span><span class="code-comment">// when <span class="code-keyword">this</span> method is called, not when we pass it to parallel.
</span><span class="code-comment">// To <span class="code-keyword">do</span> <span class="code-keyword">this</span>, you need to wrap the code below in { }, and either <span class="code-keyword">return</span>
</span><span class="code-comment">// that explicitly, or use { -> } syntax.
</span>env.inputIndex = inputIndex
<span class="code-keyword">return</span> {
stage(stepName){
powershell <span class="code-quote">"exit 1"</span>
}
}
}</pre>

---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kudiaborm.jpg" title="kudiaborm's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kudiaborm">kudiaborm</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-45277" target="_blank">Powershell Pipeline Step Hangs</a></i></summary>
<i><ul>
<li><b>status</b>: Resolved
<li><b>priority</b>: Minor
<li><b>component(s)</b>: durable-task-plugin
<li><b>resolution</b>: Won't Fix
<li><b>resolved</b>: 2017-11-21T14:42:52+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 2
<li><b>imported</b>: 2025-12-09
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p><del>When running powershell commands in parallel the stages will hang. They say <span class="error">&#91;running powershell script&#93;</span> but nothing runs, I've tried different code snippets so it doesn't appear to be related to what I'm running</del></p>

<p>Powershell pipeline task just hangs regardless of running in parallel or not. I'm running Powershell 2.0 on Windows 7</p>

<p>Job attached</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java"><span class="code-comment">//Source: https://github.com/jenkinsci/pipeline-examples/blob/master/pipeline-examples/parallel-from-list/parallelFromList.groovy
</span>
<span class="code-comment">// While you can<span class="code-quote">'t use Groovy'</span>s .collect or similar methods currently, you can
</span><span class="code-comment">// still transform a list into a set of actual build steps to be executed in
</span><span class="code-comment">// parallel.
</span>
<span class="code-comment">// Our initial list of strings we want to echo in parallel
</span>def computers = [<span class="code-quote">"COMPUTERNAME"</span>]

env.computers = computers

<span class="code-comment">// The map we'll store the parallel steps in before executing them.
</span>def stepsForParallel = [:]

<span class="code-comment">// The standard <span class="code-quote">'<span class="code-keyword">for</span> (<span class="code-object">String</span> s: stringsToEcho)'</span> syntax also doesn't work, so we
</span><span class="code-comment">// need to use old school <span class="code-quote">'<span class="code-keyword">for</span> (<span class="code-object">int</span> i = 0...)'</span> style <span class="code-keyword">for</span> loops.
</span><span class="code-keyword">for</span> (<span class="code-object">int</span> i = 0; i &lt; computers.size(); i++) {
<span class="code-comment">// Get the actual string here.
</span>def s = computers.get(i)

<span class="code-comment">// Transform that into a step and add the step to the map as the value, with
</span><span class="code-comment">// a name <span class="code-keyword">for</span> the parallel step as the key. Here, we'll just use something
</span><span class="code-comment">// like <span class="code-quote">"echoing (string)"</span>
</span>def stepName = <span class="code-quote">"Pushing to ${s}"</span>

stepsForParallel[stepName] = transformIntoStep(stepName, i)
}

<span class="code-comment">// Actually run the steps in parallel - parallel takes a map as an argument,
</span><span class="code-comment">// hence the above.
</span>node{
parallel stepsForParallel
}

<span class="code-comment">// Take the string and echo it.
</span>def transformIntoStep(stepName, inputIndex) {
<span class="code-comment">// We need to wrap what we <span class="code-keyword">return</span> in a Groovy closure, or <span class="code-keyword">else</span> it's invoked
</span><span class="code-comment">// when <span class="code-keyword">this</span> method is called, not when we pass it to parallel.
</span><span class="code-comment">// To <span class="code-keyword">do</span> <span class="code-keyword">this</span>, you need to wrap the code below in { }, and either <span class="code-keyword">return</span>
</span><span class="code-comment">// that explicitly, or use { -&gt; } syntax.
</span>env.inputIndex = inputIndex
<span class="code-keyword">return</span> {
stage(stepName){
powershell <span class="code-quote">"exit 1"</span>
}
}
}</pre>
</div></div></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Windows 7<br/>
Jenkins 2.47<br/>
Pipeline Nodes and Processes 2.12<br/>
Powershell version 2.0
```
</details>
